pyo3-macros-backend/src/pyclass.rs

```diff
@@ -628,6 +759,201 @@ fn impl_enum(
     })
 }
 
+fn impl_complex_enum(
+    complex_enum: PyClassComplexEnum<'_>,
+    args: &PyClassArgs,
+    doc: PythonDoc,
+    methods_type: PyClassMethodsType,
+) -> Result<TokenStream> {
+    // Need to rig the enum PyClass options
+    let args = {
+        let mut rigged_args = args.clone();
+        // Needs to be frozen to disallow `&mut self` methods, which could break a runtime invariant
+        rigged_args.options.frozen = parse_quote!(frozen);
```

Comment:
@davidhewitt 

I've put this in to see if it works.
It does, but the error message might be a little difficult for a user to understand.
```
error[E0271]: type mismatch resolving `<ComplexEnum as PyClass>::Frozen == False`
  --> lib.rs:48:30
   |
48 |     pub fn do_stuff_in_place(&mut self) {
   |                              ^ expected `False`, found `True`
   |
note: required by a bound in `extract_pyclass_ref_mut`
  --> /home/mate/prog/pyo3/src/impl_/extract_argument.rs:57:56
   |
57 | pub fn extract_pyclass_ref_mut<'a, 'py: 'a, T: PyClass<Frozen = False>>(
   |                                                        ^^^^^^^^^^^^^^ required by this bound in `extract_pyclass_ref_mut`
```

Do you see a way to improve this?

Reply:
I think this is totally fine; it's the same as what we already have for the rest of frozen pyclass UX (regardless of enums). It's purely a documentation problem at this point I think, outside the scope of this PR.

Reply:
I've added a test case using `test_compile_error` to document that `&mut self` methods are not supported.